src/test/java/com/google/devtools/build/lib/runtime/CommandLinePathFactoryTest.java

```diff
@@ -0,0 +1,58 @@
+package com.google.devtools.build.lib.runtime;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.assertThrows;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.devtools.build.lib.vfs.PathFragment;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests for {@link CommandLinePathFactory}. */
+@RunWith(JUnit4.class)
+public class CommandLinePathFactoryTest {
```

Comment:
Is this class also going to be used in the PATH lookup case? (i.e., `--credential_helper=somehelper`)? If so, we should have tests here for paths that don't start with a slash, or without slashes at all.

Reply:
Sorry, let me clarify:

- If we're going to use this class for PATH lookup (e.g. `--credential_helper=foo`), we should have a test for them.
- Since we agreed in the proposal to disallow non-absolute paths *with* slashes (e.g. `--credential_helper=foo/bar`) unless they're relative to a `%root%`, we should reject them (and have a test to verify that we do).

Reply:
Added `PATH` lookup (I was originally planning to do it somewhere else, but we may as well do it in this factory)

Reply:
Actually, I'm not so sure PATH lookup belongs here. (What I had in mind was to simply store the non-absolute path, and resolve it later.)

I expect this code will run once at options parsing time, which means the result will be cached during the lifetime of the Bazel server process. Hence, if you add or remove a binary from one of the PATH directories that could affect the resolution of a particular helper, it won't take effect until you restart the server, which might be surprising to users.

WDYT? Should we defer PATH resolution until the helper is invoked?

Reply:
Actually, I think what I said above only applies to startup options. If we can make `--credential_helper` a non-startup option, then I think it's fine to do the resolution here.

Reply:
Ack. I think `query` actually implies `build` for historical reasons.